lib/internal/streams/readable.js

```diff
@@ -1287,6 +1361,232 @@ ObjectDefineProperties(ReadableState.prototype, {
+          consumeFinish(this[kConsume], new AbortError());
+        }
+      });
+
+    process.nextTick(consumeStart, parent[kConsume]);
+  });
+}
+
+function consumeStart(consume) {
+  const { _readableState: state } = consume.parent;
+
+  for (const chunk of state.buffer) {
+    consumePush(consume, chunk);
+  }
+
+  if (state.endEmitted) {
+    consumeEnd(this[kConsume]);
+  } else {
+    consume.parent.once('end', function() {
+      consumeEnd(this[kConsume]);
+    });
+  }
+
+  if (consume.parent.isPaused()) {
+    consume.parent.resume();
+  }
+
+  while (true) {
+    const chunk = consume.parent.read();
+    if (chunk === null) {
+      break;
+    }
+  }
+}
+
+function consumeEnd(consume) {
+  const { type, body, resolve, decoder, parent } = consume;
+
+  try {
+    if (type === kTextType) {
+      resolve(body + (decoder ? decoder.end() : ''));
+    } else if (type === kJSONType) {
+      resolve(JSONParse(body + (decoder ? decoder.end() : '')));
+    } else if (type === kArrayBufferType) {
+      let len = 0;
+      for (const buf of body) {
+        len += buf.byteLength;
+      }
+
+      const dst = new Uint8Array(len);
+
+      let pos = 0;
+      for (const buf of body) {
+        TypedArrayPrototypeSet(dst, buf, pos);
+        pos += buf.byteLength;
+      }
+
+      resolve(dst);
+    } else if (type === kBlobType) {
+      if (!Blob) {
+        Blob = require('buffer').Blob;
+      }
+      resolve(new Blob(body));
+    }
+
+    consumeFinish(consume, undefined);
+  } catch (err) {
+    parent.destroy(err);
+  }
+}
+
+function consumePush(consume, chunk, encoding) {
+  if (!consume) {
+    return true;
+  }
+
+  if (consume.ended) {
+    return false;
+  }
+
+  if (chunk === null) {
+    consume.ended = true;
+    consume.parent.read();
+    return false;
+  }
+
+  consume.pushed = true;
+
+  if (consume.type === kTextType || consume.type === kJSONType) {
+    consumePushString(consume, chunk, encoding);
+  } else {
+    consumePushBuffer(consume, chunk, encoding);
+  }
+
+  if (!consume.parent[kReading] && !consume.reading) {
+    consume.reading = true;
+    process.nextTick(consumeReadMore, consume);
+  }
+
+  return true;
+}
+
+function consumePushString(consume, chunk, encoding) {
+  if (!consume.body) {
+    consume.body = '';
+  }
+
+  if (typeof chunk === 'string') {
+    if (consume.decoder) {
+      chunk = consume.decoder.write(Buffer.from(chunk, encoding));
+    } else if (encoding !== 'utf8') {
+      chunk = Buffer.from(chunk, encoding).toString();
+    }
+  } else if (ArrayBufferIsView(chunk)) {
+    if (!consume.decoder) {
+      consume.decoder = new StringDecoder('utf8');
+    }
+    encoding = consume.decoder.write(chunk);
+  } else {
+    throw new ERR_INVALID_ARG_TYPE(
+      'chunk', ['Buffer', 'string', 'TypedArray'], chunk);
+  }
+
+  consume.body += chunk;
+}
+
+function consumePushBuffer(consume, chunk, encoding) {
+  if (!consume.body) {
+    consume.body = [];
```

Comment:
If you collect the length of chunks on consume and increment it whenever a chunk is pushed by its length you can avoid the extra iteration in `.arrayBufer()` at the start to get the length